[real_tensor_prop] Infer Fake kernels during real tensor prop

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139213
* #139212

This PR changes real_tensor_prop to also infer fake kernels when the
operator doesn't have it.

We infer the fake output to be of the same properties as the real
output, with unbacked symints in the sizes and some stride order.

Test Plan:
- new tests

zou3519 committed:
[real_tensor_prop] Infer Fake kernels during real tensor prop

We do this if there isn't already a Fake kernel available.
We infer the fake output to be of the same properties as the real
output, with unbacked symints in the sizes and some stride order.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[real_tensor_prop] Infer Fake kernels during real tensor prop"


This PR changes real_tensor_prop to also infer fake kernels when the
operator doesn't have it.

We infer the fake output to be of the same properties as the real
output, with unbacked symints in the sizes and some stride order.

Test Plan:
- new tests

[ghstack-poisoned]

pianpwk commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -2589,3 +2599,66 @@ def dump_cache_stats() -> None:
+        raise RuntimeError(
+            f"propagate_real_tensors: we don't support operators that return "
+            f"non-Tensors. Got {op._name}"
+        )
```
could we just return the primitive values in this case? Besides tensors & those, is it just ScriptObjects that we'd have to deal with?

zou3519 replied:
If an operator returns a non-symbolic float or an int, we'd end up baking it into the graph. So the graph isn't useful for anything beyond the current sample inputs.

Also, most operators do not return non-symbolic floats or ints. If the data in the future ends up showing we need to relax this, we can. But I intended this PR to be an initial conservative version that we can expand on.

pianpwk commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -2589,3 +2599,66 @@ def dump_cache_stats() -> None:
+                )
+            fake_strides[idx] = fake_stride
+            expected = expected * real_out.shape[idx]
+            fake_stride = fake_stride * fake_shape[idx]
```
just for my understanding, why do we do need to have the `.sort()` call above, and why can't we just return the real tensor strides as-is, even if they're not dense in memory?

zou3519 replied:
>  why do we do need to have the .sort() call above

We're using this to check the tensor is dense in memory. If the tensor has sizes (a, b, c, d), then it is contiguous if the strides look like (b * c * d, c * d, d, 1). The tensor is dense in memory if the tensor can be permuted into being contiguous. We sort the strides to check if they follow that pattern

pianpwk replied:
this makes sense, thank you!

zou3519 commented:
Btw, tests are failing because there's a bug somewhere that in some of the export configurations, propagate_real_tensors attaches no real_tensors to the FakeTensors of the graph inputs. I'm working on minimizing the example and figuring that out

zou3519 committed:
Update on "[real_tensor_prop] Infer Fake kernels during real tensor prop"


This PR changes real_tensor_prop to also infer fake kernels when the
operator doesn't have it.

We infer the fake output to be of the same properties as the real
output, with unbacked symints in the sizes and some stride order.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[real_tensor_prop] Infer Fake kernels during real tensor prop"


This PR changes real_tensor_prop to also infer fake kernels when the
operator doesn't have it.

We infer the fake output to be of the same properties as the real
output, with unbacked symints in the sizes and some stride order.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge

zou3519 commented:
@pytorchbot merge -f "infinite queue"
